c_src/enif_jq.c

```diff
@@ -190,17 +439,106 @@ static ERL_NIF_TERM parse_nif(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[
 
 static ErlNifFunc nif_funcs[] = {
     /*
-       The parse_nif function seems to be very slow.
+       The parse_nif function seems to be very slow (at least when
+       given filter programs that are not in the cache).
        The Erlang VM acts very strangely when it is not scheduled
        on a dirty scheduler (for example, timer:sleep() suspends
        for a much longer time than is should).
     */
     {"parse", 2, parse_nif, ERL_NIF_DIRTY_JOB_CPU_BOUND}
 };
 
-int upgrade(ErlNifEnv* env, void** priv_data, void** old_priv_data, ERL_NIF_TERM load_info)
-{
-	return 0;
+static int get_int_config(
+        ErlNifEnv* caller_env,
+        ERL_NIF_TERM load_info,
+        char* property_name,
+        int* res_wb) {
+
+    ERL_NIF_TERM property_atom = enif_make_atom(
+            caller_env,
+            property_name);
+    ERL_NIF_TERM property_term; 
+    if ( ! enif_get_map_value(
+            caller_env,
+            load_info,
+            property_atom,
+            &property_term)) {
+        // Incorrect load info
+        return 1;
+    }
+    if ( ! enif_get_int(
+               caller_env,
+               property_term,
+               res_wb)) {
+        // Incorrect load info
+        return 1;
+    }
+    return 0;
+}
+
+static int load_helper(
+        ErlNifEnv* caller_env,
+        void** priv_data,
+        ERL_NIF_TERM load_info,
+        int nr_of_loads_before) {
+
+    int filter_program_lru_cache_max_size;
+    if ( get_int_config(
+                caller_env,
+                load_info,
+                "filter_program_lru_cache_max_size",
+                &filter_program_lru_cache_max_size) ) {
+        return 1;
+    }
+    module_private_data* data = enif_alloc(sizeof(module_private_data));
+    data->nr_of_loads_before = data->nr_of_loads_before + 1;
+    data->version = 0;
+    data->lru_cache_max_size = filter_program_lru_cache_max_size;
+    if (thrd_success !=
+        tss_create(&data->thread_local_jq_state_lru_cache_key, NULL)) {
+        enif_free(data);
+        return 1;
+    }
+    char buffer[128];
+    sprintf(buffer, "jq.module_private_data_v%d", nr_of_loads_before);
+    data->lock = enif_mutex_create(buffer);
+    if (data->lock == NULL) {
+        tss_delete(data->thread_local_jq_state_lru_cache_key);
+        enif_free(data);
+        return 1;
+    }
+    JQStateCacheEntry_lru_ptr_dynarr_init(&data->caches);
+    *priv_data = data;
+    return 0;
+}
+
+static int load(ErlNifEnv* caller_env, void** priv_data, ERL_NIF_TERM load_info) {
+    return load_helper(caller_env, priv_data, load_info, 0);
 }
 
-ERL_NIF_INIT(jq, nif_funcs, NULL, NULL, upgrade, NULL)
+void unload(ErlNifEnv* caller_env, void* priv_data) {
+    module_private_data* data = priv_data; 
```

Comment:
This can't happen if I have understood it correctly. Unload will only be called if the corresponding Erlang module is about to be removed from the system which can't happen if any process is is still holding refernces to code in the modeule. I will check again to be sure, but if that is not the case I think the design is quite strange.

Reply:
I agree.